File: Project_and_Analysis_of_Algorithms/graph_search_applications/c++/dfs.cpp
#include "graphs.h"

// initVisited maps all nodes as unvisited.
visitedMap(string) initVisited(adjacencyMap(string) graph) {
    visitedMap(string) visited;
    for (auto const&[node, neighbous] : graph) {
        visited[node] = false;
    }
    return visited;
}

// Iterative depth first search from the starting node using a queue.
void dfs(adjacencyMap(string) graph, string startNode, visitedMap(string) visited) {
    stack<string> q;
    q.push(startNode);
    visited[startNode] = true;

    while(!q.empty()) {
        string curNode = q.top(); q.pop();
        cout << curNode << "\n"; // Process Node.
        for (string neighbour : graph[curNode]) {
            if (!visited[neighbour]){
                q.push(neighbour);
                visited[neighbour] = true;
            }
        }
    }
}

int main(void) {
    dfs(G1, "A", initVisited(G1));
    return EXIT_SUCCESS;
}

File: Project_and_Analysis_of_Algorithms/graph_search_applications/python/dfs.py
"""Depth First Search"""

from graphs import G1

def init_visited(graph: list[str, list[str]]) -> dict[str, bool]:
    """Maps all nodes as unvisited."""
    return {node: False for node in graph.keys()}


def dfs(graph: list[str, list[str]], start_node: str, visited: dict[str, bool]):
    """Iterative Depth First Search starting from the start node using a queue."""
    stack = [start_node]
    visited[start_node] = True

    while len(stack) > 0:
        cur_node: str = stack.pop(-1)
        print(cur_node) # Process.
        for neighbour in graph[cur_node]:
            if not visited[neighbour]:
                stack.append(neighbour)
                visited[neighbour] = True


def main():
    dfs(graph=G1, start_node="A", visited=init_visited(G1))


if __name__ == "__main__":
    main()
